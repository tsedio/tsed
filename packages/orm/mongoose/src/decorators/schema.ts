import {decoratorTypeOf, StoreMerge, useDecorators} from "@tsed/core";
import {registerProvider} from "@tsed/di";
import {Property} from "@tsed/schema";
import {SchemaTypeOptions} from "mongoose";
import {MONGOOSE_SCHEMA} from "../constants";
import {MongooseSchemaOptions} from "../interfaces";
import {getSchema, getSchemaToken} from "../utils/createSchema";

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function Schema(options?: MongooseSchemaOptions): (target: any) => void;
export function Schema(definition: SchemaTypeOptions<any>): Function;
export function Schema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return (...parameters: any[]) => {
    switch (decoratorTypeOf(parameters)) {
      case "property":
        return useDecorators(Property(), StoreMerge(MONGOOSE_SCHEMA, options))(parameters[0], parameters[1], parameters[2]);

      case "class":
        const {token} = getSchemaToken(parameters[0], options);

        registerProvider({
          provide: token,
          deps: [],
          useFactory() {
            return getSchema(parameters[0], options as any);
          }
        });
        break;
    }
  };
}

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function MongooseSchema(options?: MongooseSchemaOptions): (target: any) => void;
export function MongooseSchema(definition: SchemaTypeOptions<any>): Function;
export function MongooseSchema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return Schema(options as any);
}
