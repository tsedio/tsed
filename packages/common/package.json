{
  "name": "@tsed/common",
  "version": "6.63.2",
  "description": "A TypeScript Framework on top of Express",
  "source": "./src/index.ts",
  "main": "./lib/index.js",
  "typings": "./lib/index.d.ts",
  "private": false,
  "keywords": [
    "Express",
    "Koa",
    "TypeScript",
    "typescript",
    "Decorator",
    "decorators",
    "decorator",
    "express",
    "koa",
    "Controller",
    "Inject",
    "ioc",
    "di",
    "mvc",
    "swagger",
    "swagger ui",
    "ES2015",
    "ES6",
    "server",
    "rest",
    "api",
    "validation"
  ],
  "author": {
    "name": "Romain Lenzotti"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tsedio/tsed/issues"
  },
  "homepage": "http://tsed.io/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tsedio//tsed.git"
  },
  "scripts": {
    "build": "tsc --build tsconfig.compile.json"
  },
  "dependencies": {
    "@tsed/core": "6.63.2",
    "@tsed/perf": "6.63.2",
    "@tsed/di": "6.63.2",
    "@tsed/exceptions": "6.63.2",
    "@tsed/json-mapper": "6.63.2",
    "@tsed/logger": "^5.16.0",
    "@tsed/schema": "6.63.2",
    "@types/cache-manager": "^3.4.0",
    "@types/json-schema": "7.0.7",
    "cache-manager": "^3.4.1",
    "consolidate": "^0.16.0",
    "ejs": "^3.1.5",
    "json-schema": "^0.2.3",
    "on-finished": "2.3.0",
    "tslib": "2.2.0",
    "uuid": "8.3.2"
  },
  "devDependencies": {
    "@tsed/engines": "^1.1.4",
    "@types/consolidate": "0.14.0",
    "rxjs": "^6.5.2"
  },
  "peerDependencies": {
    "@tsed/engines": "^1.1.4"
  }
}