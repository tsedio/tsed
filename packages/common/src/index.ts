export * from "@tsed/di";
export * from "@tsed/platform-exceptions";
export * from "@tsed/platform-response-filter";
export * from "@tsed/platform-cache";
export * from "./config";
export * from "./mvc";
export * from "./platform";
export * from "./platform-test";

export {$log, Logger} from "@tsed/logger";
