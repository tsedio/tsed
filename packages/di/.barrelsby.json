{
  "directory": ["./src/common", "./src/node"],
  "exclude": ["**/__mock__", "**/__mocks__", "**/*.spec.ts"],
  "delete": true
}
