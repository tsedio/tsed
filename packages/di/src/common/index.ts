/**
 * @file Automatically generated by @tsed/barrels.
 */
export * from "./constants/constants.js";
export * from "./decorators/autoInjectable.js";
export * from "./decorators/configuration.js";
export * from "./decorators/constant.js";
export * from "./decorators/controller.js";
export * from "./decorators/inject.js";
export * from "./decorators/injectable.js";
export * from "./decorators/intercept.js";
export * from "./decorators/interceptor.js";
export * from "./decorators/lazyInject.js";
export * from "./decorators/module.js";
export * from "./decorators/opts.js";
export * from "./decorators/overrideProvider.js";
export * from "./decorators/scope.js";
export * from "./decorators/service.js";
export * from "./decorators/useOpts.js";
export * from "./decorators/value.js";
export * from "./domain/Container.js";
export * from "./domain/ControllerProvider.js";
export * from "./domain/InjectablePropertyType.js";
export * from "./domain/LocalsContainer.js";
export * from "./domain/Provider.js";
export * from "./domain/ProviderScope.js";
export * from "./domain/ProviderType.js";
export * from "./errors/InjectionError.js";
export * from "./errors/InvalidPropertyTokenError.js";
export * from "./errors/UndefinedTokenError.js";
export * from "./interfaces/DIConfigurationOptions.js";
export * from "./interfaces/DILogger.js";
export * from "./interfaces/DILoggerOptions.js";
export * from "./interfaces/DIResolver.js";
export * from "./interfaces/ImportTokenProviderOpts.js";
export * from "./interfaces/InjectableProperties.js";
export * from "./interfaces/InterceptorContext.js";
export * from "./interfaces/InterceptorMethods.js";
export * from "./interfaces/InvokeOptions.js";
export * from "./interfaces/OnDestroy.js";
export * from "./interfaces/OnInit.js";
export * from "./interfaces/ProviderOpts.js";
export * from "./interfaces/RegistrySettings.js";
export * from "./interfaces/ResolvedInvokeOptions.js";
export * from "./interfaces/TokenProvider.js";
export * from "./interfaces/TokenRoute.js";
export * from "./registries/GlobalProviders.js";
export * from "./registries/ProviderRegistry.js";
export * from "./services/DIConfiguration.js";
export * from "./services/DILogger.js";
export * from "./services/InjectorService.js";
export * from "./utils/colors.js";
export * from "./utils/createContainer.js";
export * from "./utils/getConfiguration.js";
export * from "./utils/getConstructorDependencies.js";
export * from "./utils/resolveControllers.js";
