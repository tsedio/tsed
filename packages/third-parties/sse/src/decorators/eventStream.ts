import {StoreSet, useDecorators} from "@tsed/core";
import {Use} from "@tsed/platform-middlewares";
import {Header, AcceptMime} from "@tsed/schema";
import {EventStreamMiddleware} from "../middlewares/EventStreamMiddleware.js";

export interface EventStreamOpts {
  headers?: Record<string, string>;
  event?: string;
}

export function EventStream(opts: EventStreamOpts = {}) {
  return useDecorators(
    StoreSet(EventStreamMiddleware, opts),
    Use(EventStreamMiddleware),
    AcceptMime("text/event-stream"),
    Header("Content-Type", "text/event-stream")
  );
}
