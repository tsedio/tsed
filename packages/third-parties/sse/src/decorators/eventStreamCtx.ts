import {Context} from "@tsed/platform-params";

import type {EventStreamContext} from "../domain/EventStreamContext.js";

export type EventStreamCtx = EventStreamContext;

export function EventStreamCtx() {
  return Context("EventStreamContext");
}
