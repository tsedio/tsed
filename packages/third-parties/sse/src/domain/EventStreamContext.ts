import {PlatformContext} from "@tsed/common";
import {isBoolean, isDate, isNumber, isString} from "@tsed/core";
import {PlatformResponseFilter} from "@tsed/platform-response-filter";
import type {EventEmitter} from "events";
import type {Observable} from "rxjs";

import type {EventStreamOpts} from "../decorators/eventStream.js";

export type EventStreamContextOptions = {
  $ctx: PlatformContext;
} & EventStreamOpts;

export class EventStreamContext {
  private $ctx: PlatformContext;
  private opts: EventStreamOpts;
  private responseFilter: PlatformResponseFilter;
  #isDone = false;

  constructor({$ctx, ...opts}: EventStreamContextOptions) {
    this.$ctx = $ctx;

    this.opts = opts;
    this.responseFilter = this.$ctx.injector.get<PlatformResponseFilter>(PlatformResponseFilter)!;
  }

  writeHead(headers: Record<string, string> = {}) {
    console.log("this.$ctx", this.$ctx);
    this.on("end", () => {
      this.$ctx.logger.info({
        event: "SSE_ENDED"
      });
      this.$ctx.getRes().end();
      this.#isDone = true;
    });

    this.on("close", () => {
      this.$ctx.logger.info({
        event: "SSE_CLOSED_BY_CLIENT"
      });
      this.$ctx.getRes().end();
      this.#isDone = true;
    });

    this.$ctx.getRes().writeHead(200, {
      ...this.opts.headers,
      ...headers,
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      Connection: "keep-alive"
    });
  }

  on(event: string, listener: (...args: any[]) => void): this {
    this.$ctx.getRes().on(event, listener);
    return this;
  }

  subscribe(observable: Observable<unknown>) {
    observable.subscribe((data) => {
      this.emit(this.opts.event, data);
    });
    return this;
  }

  stream(stream: NodeJS.ReadableStream | EventEmitter) {
    const listener = (data: unknown) => {
      if (this.isDone()) {
        stream.off("data", listener);
      } else {
        this.emit(this.opts.event, data);
      }
    };

    stream.on("data", listener);
  }

  emit(event: string | undefined, data: unknown) {
    if (this.isDone()) {
      return;
    }
    const {$ctx} = this;
    if (data === undefined) {
      return this;
    }

    if (Buffer.isBuffer(data)) {
      data = data.toString("utf8");
    }

    if (isBoolean(data) || isNumber(data) || isString(data) || isDate(data) || data === null) {
      this.write({
        event: event || "",
        data: JSON.stringify(data)
      });
      return this;
    }

    this.responseFilter
      .serialize(data, $ctx as any)
      .then((data: unknown) => {
        return this.responseFilter.transform(data, $ctx as any);
      })
      .then((data: unknown) => {
        data = JSON.stringify(data);

        this.write({
          event: event || "",
          data: data as string
        });
      });

    return this;
  }

  write(obj: Record<string, string>) {
    const values = Object.entries(obj)
      .map(([key, value]) => {
        return `${key}: ${value}`;
      })
      .join("\n");

    this.$ctx.getRes().write(`${values}\n\n`);
  }

  isDone() {
    return this.#isDone;
  }

  end() {
    this.#isDone = true;
    this.$ctx.getRes().end();
  }
}
