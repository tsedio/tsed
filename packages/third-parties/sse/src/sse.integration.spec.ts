import {Context, PlatformContext, PlatformTest} from "@tsed/common";
import {Controller} from "@tsed/di";
import {Get, getSpec, JsonEntityStore, Name, Returns} from "@tsed/schema";

import {EventStream} from "./decorators/eventStream.js";
import {EventStreamCtx} from "./decorators/eventStreamCtx.js";
import {EventStreamContext} from "./domain/EventStreamContext.js";

function make(eventStream: EventStreamCtx, cb: () => any) {
  let intervalId: ReturnType<typeof setInterval>;

  eventStream.on("close", () => {
    clearInterval(intervalId);
  });

  eventStream.on("end", () => {
    clearInterval(intervalId);
  });

  intervalId = setInterval(() => {
    // Ts.ED support Model serialization using json-mapper here
    eventStream.emit("event", cb());
  }, 200);
}

class Model {
  @Name("id")
  _id: string;
}

@Controller("/sse")
export class MyCtrl {
  @Get("/scenario1")
  @EventStream()
  scenario1(@EventStreamCtx() eventStream: EventStreamCtx) {
    make(eventStream, () => new Date());
  }

  @Get("/scenario2")
  @EventStream()
  scenario2(@EventStreamCtx() eventStream: EventStreamCtx) {
    make(eventStream, () => false);
  }

  @Get("/scenario3")
  @EventStream()
  scenario3(@EventStreamCtx() eventStream: EventStreamCtx) {
    make(eventStream, () => 0);
  }

  @Get("/scenario4")
  @EventStream()
  scenario4(@EventStreamCtx() eventStream: EventStreamCtx) {
    make(eventStream, () => ({
      date: new Date(),
      message: "Hello"
    }));
  }

  @Get("/scenario5")
  @EventStream()
  @(Returns(200).Type(Model))
  scenario5(@EventStreamCtx() eventStream: EventStreamCtx, @Context() $ctx: PlatformContext) {
    make(eventStream, () => {
      const model = new Model();
      model._id = new Date().toISOString();

      return model;
    });
  }
}

async function getControllerFixture(scenario: string) {
  const controller = await PlatformTest.invoke<MyCtrl>(MyCtrl);

  const endpoint = JsonEntityStore.fromMethod(MyCtrl, scenario);
  const $ctx = PlatformTest.createRequestContext({
    endpoint
  });
  const eventStream = new EventStreamContext({
    $ctx
  });

  return {controller, endpoint, eventStream, $ctx};
}

describe("SSE integration", () => {
  beforeEach(() => PlatformTest.create());
  afterEach(() => PlatformTest.reset());
  beforeEach(() => {
    vi.useFakeTimers({
      toFake: ["Date"],
      now: new Date("2025-01-01T00:00:00Z").getTime()
    });
  });
  afterEach(() => {
    vi.useRealTimers();
  });
  it("should generate the swagger spec", () => {
    expect(getSpec(MyCtrl)).toMatchSnapshot();
  });

  describe("Scenario1 - Date/String", () => {
    it("should pass the new EventStreamContext to function and process message", async () => {
      const {controller, eventStream} = await getControllerFixture("scenario1");

      controller.scenario1(eventStream);

      // Wait for the first message
      const result = await new Promise((resolve) => {
        eventStream.on("data", resolve);
      });

      expect(result).toEqual("event: event\ndata: " + JSON.stringify(new Date()) + "\n\n");

      // stop the interval
      eventStream.end();
    });
  });

  describe("Scenario2 - Boolean", () => {
    it("should pass the new EventStreamContext to function and process message", async () => {
      const {controller, eventStream} = await getControllerFixture("scenario2");

      controller.scenario2(eventStream);

      // Wait for the first message
      const result = await new Promise((resolve) => {
        eventStream.on("data", resolve);
      });

      expect(result).toEqual("event: event\ndata: false\n\n");

      // stop the interval
      eventStream.end();
    });
  });

  describe("Scenario3 - Number", () => {
    it("should pass the new EventStreamContext to function and process message", async () => {
      const {controller, eventStream} = await getControllerFixture("scenario3");

      controller.scenario3(eventStream);

      // Wait for the first message
      const result = await new Promise((resolve) => {
        eventStream.on("data", resolve);
      });

      expect(result).toEqual("event: event\ndata: 0\n\n");

      // stop the interval
      eventStream.end();
    });
  });

  describe("Scenario4 - Object", () => {
    it("should pass the new EventStreamContext to function and process message", async () => {
      const {controller, eventStream} = await getControllerFixture("scenario4");

      controller.scenario4(eventStream);

      // Wait for the first message
      const result = await new Promise((resolve) => {
        eventStream.on("data", resolve);
      });

      expect(result).toEqual(
        "event: event\ndata: " +
          JSON.stringify({
            date: "2025-01-01T00:00:00.000Z",
            message: "Hello"
          }) +
          "\n\n"
      );

      // stop the interval
      eventStream.end();
    });
  });

  describe("Scenario5 - Model", () => {
    it("should pass the new EventStreamContext to function and process message", async () => {
      const {controller, eventStream, $ctx} = await getControllerFixture("scenario5");

      controller.scenario5(eventStream, $ctx);

      // Wait for the first message
      const result = await new Promise((resolve) => {
        eventStream.on("data", resolve);
      });

      expect(result).toEqual(
        "event: event\ndata: " +
          JSON.stringify({
            id: "2025-01-01T00:00:00.000Z"
          }) +
          "\n\n"
      );

      // stop the interval
      eventStream.end();
    });
  });
});
