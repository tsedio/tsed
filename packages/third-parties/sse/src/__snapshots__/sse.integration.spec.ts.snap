// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SSE integration > should generate the swagger spec 1`] = `
{
  "components": {
    "schemas": {
      "Model": {
        "properties": {
          "id": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "paths": {
    "/sse/scenario1": {
      "get": {
        "operationId": "myCtrlScenario1",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "*/*": {
                "schema": {
                  "type": "object",
                },
              },
            },
            "headers": {
              "Content-Type": {
                "example": "text/event-stream",
                "schema": {
                  "type": "string",
                },
              },
            },
          },
        },
        "tags": [
          "MyCtrl",
        ],
      },
    },
    "/sse/scenario2": {
      "get": {
        "operationId": "myCtrlScenario2",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "*/*": {
                "schema": {
                  "type": "object",
                },
              },
            },
            "headers": {
              "Content-Type": {
                "example": "text/event-stream",
                "schema": {
                  "type": "string",
                },
              },
            },
          },
        },
        "tags": [
          "MyCtrl",
        ],
      },
    },
    "/sse/scenario3": {
      "get": {
        "operationId": "myCtrlScenario3",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "*/*": {
                "schema": {
                  "type": "object",
                },
              },
            },
            "headers": {
              "Content-Type": {
                "example": "text/event-stream",
                "schema": {
                  "type": "string",
                },
              },
            },
          },
        },
        "tags": [
          "MyCtrl",
        ],
      },
    },
    "/sse/scenario4": {
      "get": {
        "operationId": "myCtrlScenario4",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "*/*": {
                "schema": {
                  "type": "object",
                },
              },
            },
            "headers": {
              "Content-Type": {
                "example": "text/event-stream",
                "schema": {
                  "type": "string",
                },
              },
            },
          },
        },
        "tags": [
          "MyCtrl",
        ],
      },
    },
    "/sse/scenario5": {
      "get": {
        "operationId": "myCtrlScenario5",
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "*/*": {
                "schema": {
                  "type": "object",
                },
              },
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Model",
                },
              },
            },
            "description": "Success",
            "headers": {
              "Content-Type": {
                "example": "text/event-stream",
                "schema": {
                  "type": "string",
                },
              },
            },
          },
        },
        "tags": [
          "MyCtrl",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "MyCtrl",
    },
  ],
}
`;
